src/devices/src/virtio/block/io/async_io.rs

```diff
@@ -180,27 +180,31 @@ impl<T> AsyncFileEngine<T> {
         self.ring.submit().map(|_| ()).map_err(Error::IoUring)
     }
 
-    pub fn drain(&mut self, flush: bool) -> Result<(), Error> {
-        self.drain_submission_queue()?;
+    pub fn drain(&mut self, flush: bool, discard: bool) -> Result<(), Error> {
+        self.drain_submission_queue(flush)?;
 
-        // Drain the completion queue so that we may deallocate the user_data fields.
-        while self.do_pop()?.is_some() {}
+        if discard {
+            // Drain the completion queue so that we may deallocate the user_data fields.
+            while self.do_pop()?.is_some() {}
+        }
+
+        Ok(())
+    }
+
+    pub fn drain_submission_queue(&mut self, flush: bool) -> Result<(), Error> {
+        self.ring
+            .submit_and_wait_all()
+            .map(|_| ())
+            .map_err(Error::IoUring)?;
 
         if flush {
             // Sync data out to physical media on host.
             // We don't need to call flush first since all the ops are performed through io_uring
             // and Rust shouldn't manage any data in its internal buffers.
-            self.file.sync_all().map_err(Error::SyncAll)
-        } else {
-            Ok(())
+            self.file.sync_all().map_err(Error::SyncAll)?;
```

Comment:
I really think we should take this flush option out of the `drain_submission_queue` function to get better separation of concerns

Reply:
If we separate flush, it will not work in any conditions. A separate flush will work correctly only after a drain. So I would keep it as it is in order to avoid the possibility of using it incorrectly.

Reply:
It's just that there is nothing in the naming of the function that suggests it also flushes the file. That's why I find it confusing.
I am however not super keen on this, it's only a minor thing

Reply:
You could have a `flush` function that does drain+flush, and this `drain()` fn only does drain.

then they can be called depending on needs and none could be incorrectly used.

Just sayin :smiley:

Reply:
Done. I defined 2 methods: `drain()` and `drain_and_flush()`.